lib/multi.c

```diff
@@ -1159,22 +1173,38 @@ static CURLMcode Curl_multi_wait(struct Curl_multi *multi,
 
       for(i = 0; i < MAX_SOCKSPEREASYHANDLE; i++) {
         curl_socket_t s = CURL_SOCKET_BAD;
-
+#ifdef USE_WINSOCK
+        long mask = 0;
+#endif
         if(bitmap & GETSOCK_READSOCK(i)) {
-          ufds[nfds].fd = sockbunch[i];
+          s = sockbunch[i];
+#ifdef USE_WINSOCK
+          mask |= FD_READ|FD_ACCEPT|FD_CLOSE;
+#endif
+          ufds[nfds].fd = s;
           ufds[nfds].events = POLLIN;
           ++nfds;
-          s = sockbunch[i];
         }
         if(bitmap & GETSOCK_WRITESOCK(i)) {
-          ufds[nfds].fd = sockbunch[i];
+          s = sockbunch[i];
+#ifdef USE_WINSOCK
+          mask |= FD_WRITE|FD_CONNECT|FD_CLOSE;
+          send(s, NULL, 0, 0); /* reset FD_WRITE */
+#endif
+          ufds[nfds].fd = s;
           ufds[nfds].events = POLLOUT;
           ++nfds;
-          s = sockbunch[i];
         }
         if(s == CURL_SOCKET_BAD) {
```

Comment:
@bagder while working on this it came to my mind if this check shouldn't be at the top of the loop to avoid counting one invalid socket?

Reply:
> shouldn't be at the top of the loop to avoid counting one invalid socket?

it seems s is only set to a valid socket if the bitmap for readable/writable was set for that socket, and so at this point we're basically checking whether that has happened, but I don't know why it's breaking if the socket is invalid I would expect it to continue checking the bitmap.

Reply:
@bagder would you mind explaining why the bitmap is not further checked if readable/writable was set for a socket?

Reply:
Resolved, see https://github.com/curl/curl/pull/6245#issuecomment-797323267.